### ISSUE TITLE
sqlite3.OperationalError: database is locked when multiple drives used

### ISSUE THREAD

**Issue by sureshjoshi** [automatic-ripping-machine/automatic-ripping-machine#311]
### Description

I've started testing out v2.1_dev, and it works very well when a single CD/DVD is ripped at a time, however, when trying to run in parallel - I run into a `sqlite3.OperationalError: database is locked`.

2 DVDs and 1 music CD at the same time.

### Environment

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"
ARM: v2.1_dev

### Log

```
[2019-10-12 01:37:28] INFO ARM: makemkv.get_track_info Found 9 titles
[2019-10-12 01:37:33] ERROR ARM: main.<module> A fatal error has occured and ARM is exiting.  See traceback below for details.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/arm/arm/ripper/main.py", line 416, in <module>
    main(logfile, job)
  File "/opt/arm/arm/ripper/main.py", line 163, in main
    mkvoutpath = makemkv.makemkv(logfile, job)
  File "/opt/arm/arm/ripper/makemkv.py", line 97, in makemkv
    get_track_info(mdisc, job)
  File "/opt/arm/arm/ripper/makemkv.py", line 247, in get_track_info
    db.session.commit()
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/scoping.py", line 162, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 1027, in commit
    self.transaction.commit()
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 494, in commit
    self._prepare_impl()
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 473, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2470, in flush
    self._flush(objects)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2608, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 153, in reraise
    raise value
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/session.py", line 2568, in _flush
    flush_context.execute()
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/unitofwork.py", line 589, in execute
    uow,
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/persistence.py", line 236, in save_obj
    update,
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/persistence.py", line 996, in _emit_update_statements
    statement, multiparams
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py", line 580, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE job SET no_of_titles=? WHERE job.job_id = ?]
[parameters: (9, 19)]
(Background on this error at: http://sqlalche.me/e/e3q8)
```

**Comment by sureshjoshi** [comment]
As far as I can tell, there are no try/except semantics around writing to the sqlite DB. SQLite is great for multi-thread/process reads, but pretty weak on the writes - without specific connection handling.

In order to avoid digging through all the code to try and fix the problem (which may involve more than just try/excepts), I'm going to see how easy it is to plug in 1 DB per DVD instance, and have those all reconcile back into the UI.

**Comment by sureshjoshi** [comment]
I committed a hacky workaround to my fork (https://github.com/sureshjoshi/automatic-ripping-machine/commit/e9cc8a75fe980fc8eca1e85024922a3ce0472e24).

Essentially, each optical drive now has it's own DB file (e.g `arm.sr0.db`) and log files (e.g. `music_cd.sr2.db`). There are no checks on the optical drive naming, so they need to be filename-safe.

The DB file modification was necessary in order to remove the multi-process SQLite locking issue, but I don't like it. I spent an hour or so trying to monkey-patch the project to safely handle locks, but then gave up and went this route.

I split off the logs for my own purposes, because it helps me isolate if any of my optical drives are consistently giving me problems.

An unfortunate consequence of this is that the handy progress web server still wants to use only a single DB file. I didn't really have time to setup a merge or bind solution, so instead, I just startup a Flask instance for each optical drive (each webpage is filtered to just that drive's data), and the ports are indexed assuming there is a numerical ending to the drive name (e.g. sr0, sr1, sr2, ... srN).

`/usr/bin/python3 /opt/arm/arm/runui.py -d sr0` available at `0.0.0.0:8080`
`/usr/bin/python3 /opt/arm/arm/runui.py -d sr1` available at `0.0.0.0:8081`
`/usr/bin/python3 /opt/arm/arm/runui.py -d sr2` available at `0.0.0.0:8082`

**Comment by alexmadsen1** [comment]
I am seeing the same sqlalchemy issue. Any progress?

[2021-03-24 21:49:01] INFO ARM: identify.identify Disc title: Children of Men : 2006 : movie
[2021-03-24 21:49:01] DEBUG ARM: identify.identify #### --- job ----Job: (_sa_instance_state=<sqlalchemy.orm.state.InstanceState object at 0x7f4f518a9790>) (hasnicetitle=True) (pid_hash=-6706668389693682299) (imdb_id=tt0206634) (title_manual=None) (logfile=CHILDREN_OF_MEN.log) (errors=None) (start_time=2021-03-24 21:48:59.799801) (year=2006) (imdb_id_auto=tt0206634) (year_auto=2006) (imdb_id_manual=None) (stop_time=None) (disctype=dvd) (job_length=None) (year_manual=None) (poster_url=https://m.media-amazon.com/images/M/MV5BMTQ5NTI2NTI4NF5BMl5BanBnXkFtZTcwNjk2NDA2OQ@@._V1_SX300.jpg) (label=CHILDREN_OF_MEN) (job_id=61) (status=active) (ejected=False) (poster_url_auto=https://m.media-amazon.com/images/M/MV5BMTQ5NTI2NTI4NF5BMl5BanBnXkFtZTcwNjk2NDA2OQ@@._V1_SX300.jpg) (video_type=movie) (arm_version=2.2.0-dev) (poster_url_manual=None) (video_type_auto=movie) (no_of_titles=None) (updated=False) (title=Children of Men) (pid=3575) (video_type_manual=None) (crc_id=520b39ad7f8a65cc) (devpath=/dev/sr1) (title_auto=Children of Men) (mountpoint=/mnt/dev/sr1) 
[2021-03-24 21:49:01] INFO ARM: main.main Waiting 10 seconds for manual override.
[2021-03-24 21:49:08] INFO ARM: main.<module> Starting ARM processing at 2021-03-24 21:49:08.548968
[2021-03-24 21:49:08] DEBUG ARM: utils.check_db_version Head is: c3a3fa694636
[2021-03-24 21:49:08] DEBUG ARM: utils.check_db_version Database version is: c3a3fa694636
[2021-03-24 21:49:08] INFO ARM: utils.check_db_version Database is up to date
[2021-03-24 21:49:08] INFO ARM: main.<module> ARM version: 2.2.0-dev
[2021-03-24 21:49:08] INFO ARM: main.<module> Python version: 3.8.6 (default, Jan 27 2021, 15:42:20) [GCC 10.2.0]
[2021-03-24 21:49:08] INFO ARM: main.<module> User is: root
[2021-03-24 21:49:08] INFO ARM: logger.cleanuplogs Looking for log files older than 7 days old.
[2021-03-24 21:49:08] INFO ARM: main.<module> Job: CHILDREN_OF_MEN
[2021-03-24 21:49:08] INFO ARM: main.<module> Job #61 with PID 3575 is currently running.
[2021-03-24 21:49:08] INFO ARM: main.<module> Job #62 with PID 3671 is currently running.
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params **** Logging udev attributes ****
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params DEVLINKS:/dev/cdrom /dev/disk/by-path/pci-0000:02:05.0-ata-1 /dev/disk/by-id/ata-VMware_Virtual_SATA_CDRW_Drive_00000000000000000001 /dev/cdrw /dev/dvd /dev/disk/by-path/pci-0000:02:05.0-ata-1.0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params DEVNAME:/dev/sr0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params DEVPATH:/devices/pci0000:00/0000:00:11.0/0000:02:05.0/ata3/host2/target2:0:0/2:0:0:0/block/sr0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params DEVTYPE:disk
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_ATA:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_ATA_FEATURE_SET_PM:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_ATA_FEATURE_SET_PM_ENABLED:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_BUS:ata
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_CD_R:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_CD_RW:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_DVD:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_DVD_R:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_DVD_RAM:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_MRW:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_CDROM_MRW_W:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_FOR_SEAT:block-pci-0000_02_05_0-ata-1_0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_MODEL:VMware_Virtual_SATA_CDRW_Drive
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_MODEL_ENC:VMware\x20Virtual\x20SATA\x20CDRW\x20Drive\x20\x20\x20\x20\x20\x20\x20\x20\x20\x20
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_PATH:pci-0000:02:05.0-ata-1.0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_PATH_ATA_COMPAT:pci-0000:02:05.0-ata-1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_PATH_TAG:pci-0000_02_05_0-ata-1_0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_REVISION:00000001
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_SERIAL:VMware_Virtual_SATA_CDRW_Drive_00000000000000000001
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_SERIAL_SHORT:00000000000000000001
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params ID_TYPE:cd
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params MAJOR:11
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params MINOR:0
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params SUBSYSTEM:block
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params SYSTEMD_MOUNT_DEVICE_BOUND:1
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params TAGS::uaccess:systemd:seat:
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params USEC_INITIALIZED:6075980
[2021-03-24 21:49:08] DEBUG ARM: main.log_udev_params **** End udev attributes ****
[2021-03-24 21:49:08] INFO ARM: main.main Starting Disc identification
[2021-03-24 21:49:08] INFO ARM: identify.identify Mounting disc to: /mnt/dev/sr1
[2021-03-24 21:50:09] DEBUG ARM: identify.identify Found: /mnt/dev/sr1/VIDEO_TS
[2021-03-24 21:50:09] INFO ARM: identify.identify Disc identified as video
[2021-03-24 21:50:09] ERROR ARM: main.<module> A fatal error has occured and ARM is exiting.  See traceback below for details.
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1705, in _execute_context
    self.dialect.do_execute(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/arm/arm/ripper/main.py", line 502, in <module>
    main(logfile, job)
  File "/opt/arm/arm/ripper/main.py", line 158, in main
    db.session.commit()
  File "<string>", line 2, in commit
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 1406, in commit
    self._transaction.commit(_to_root=self.future)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 820, in commit
    self._prepare_impl()
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 799, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 3224, in flush
    self._flush(objects)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 3364, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/util/compat.py", line 198, in raise_
    raise exception
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 3324, in _flush
    flush_context.execute()
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/unitofwork.py", line 453, in execute
    rec.execute(self)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/unitofwork.py", line 627, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/persistence.py", line 234, in save_obj
    _emit_update_statements(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/persistence.py", line 998, in _emit_update_statements
    c = connection._execute_20(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/sql/elements.py", line 313, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1389, in _execute_clauseelement
    ret = self._execute_context(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1748, in _execute_context
    self._handle_dbapi_exception(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1929, in _handle_dbapi_exception
    util.raise_(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/util/compat.py", line 198, in raise_
    raise exception
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/base.py", line 1705, in _execute_context
    self.dialect.do_execute(
  File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/engine/default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: UPDATE job SET status=? WHERE job.job_id = ?]
[parameters: ('active', 61)]
(Background on this error at: http://sqlalche.me/e/14/e3q8)
[2021-03-24 21:50:09] INFO ARM: identify.identify_dvd DVD CRC64 hash is: 520b39ad7f8a65cc
[2021-03-24 21:50:09] DEBUG ARM: identify.identify_dvd http://metaservices.windowsmedia.com/pas_dvd_B/template/GetMDRDVDByCRC.xml?CRC=520b39ad7f8a65cc
[2021-03-24 21:50:09] DEBUG ARM: identify.identify_dvd dvd_title_label= CHILDREN_OF_MEN
[2021-03-24 21:50:09] DEBUG ARM: identify.identify_dvd dvd_title ^a-z= CHILDREN OF MEN
[2021-03-24 21:50:09] DEBUG ARM: identify.identify_dvd dvd_title SKU$= CHILDREN OF MEN
[2021-03-24 21:50:09] DEBUG ARM: identify.callwebservice http://www.omdbapi.com/?t=CHILDREN OF MEN&y=&type=movie&plot=short&r=json&apikey=key_hidden
[2021-03-24 21:50:09] DEBUG ARM: identify.callwebservice ***Calling webservice with Title: CHILDREN OF MEN and Year: 
[2021-03-24 21:50:09] DEBUG ARM: identify.callwebservice Webservice failed
[2021-03-24 21:50:09] DEBUG ARM: identify.identify_dvd DVD_INFO_XML: fail

**Comment by 1337-server** [comment]
> 
> 
> I am seeing the same sqlalchemy issue. Any progress?

https://github.com/1337-server/automatic-ripping-machine/blob/610887548874c2ff5198642b007ce3e658e52581/arm/ripper/utils.py#L722


This function may help, but you would need to update all of ARM to use that instead of the db.session.commit()
This hasn't been added to the ARM repo yet.

Or you can try the solution just above your comment.

> 
> 
> I committed a hacky workaround to my fork ([sureshjoshi@e9cc8a7](https://github.com/sureshjoshi/automatic-ripping-machine/commit/e9cc8a75fe980fc8eca1e85024922a3ce0472e24)).
> 
> Essentially, each optical drive now has it's own DB file (e.g `arm.sr0.db`) and log files (e.g. `music_cd.sr2.db`). There are no checks on the optical drive naming, so they need to be filename-safe.
> 
> The DB file modification was necessary in order to remove the multi-process SQLite locking issue, but I don't like it. I spent an hour or so trying to monkey-patch the project to safely handle locks, but then gave up and went this route.
> 
> I split off the logs for my own purposes, because it helps me isolate if any of my optical drives are consistently giving me problems.
> 
> An unfortunate consequence of this is that the handy progress web server still wants to use only a single DB file. I didn't really have time to setup a merge or bind solution, so instead, I just startup a Flask instance for each optical drive (each webpage is filtered to just that drive's data), and the ports are indexed assuming there is a numerical ending to the drive name (e.g. sr0, sr1, sr2, ... srN).
> 
> `/usr/bin/python3 /opt/arm/arm/runui.py -d sr0` available at `0.0.0.0:8080`
> `/usr/bin/python3 /opt/arm/arm/runui.py -d sr1` available at `0.0.0.0:8081`
> `/usr/bin/python3 /opt/arm/arm/runui.py -d sr2` available at `0.0.0.0:8082`
